EcoSoapBank/Helpers/URLSessionDataTaskMock.swift

```diff
@@ -0,0 +1,21 @@
+//
+//  URLSessionDataTaskMock.swift
+//  EcoSoapBank
+//
+//  Created by Christopher Devito on 8/13/20.
+//  Copyright © 2020 Spencer Curtis. All rights reserved.
+//
+
+import Foundation
+
+class URLSessionDataTaskMock: URLSessionDataTask {
+    private let closure: () -> Void
+
+    init(closure: @escaping () -> Void) {
+        self.closure = closure
+    }
+
+    override func resume() {
+        closure()
+    }
+}
```

Comment:
With this and the mock URLSession, there could be potential unforeseen ramifications if you accidentally call any of the methods you don't override here. It may also be accessible by other initializers. It might be safer to use protocols all around so you only expose the APIs you need to (see below), and then have the mock session just conform to the protocol. That way you can avoid the issue of subclasses altogether. I think this is how the Lambda iOS curriculum had us do it. It may require some minor refactoring, but like I was saying to Shawn yesterday, better to make the compiler prevent you from making mistakes than just try to remember not to make them, IMO.

```swift
protocol DataLoader {
   func loadData(with request: URLRequest, completion: @escaping CompletionHandler) // returns nothing
}

extension URLSession: DataLoader {
  func loadData(with request: URLRequest, completion: @escaping CompletionHandler) {
    self.dataTask(with: request, completionHandler: completion).resume()
  }
}
```

If you still wanted it to return a data task (in case you want to `cancel` it in some cases, you could also create a `DataTask` protocol that `URLSessionDataTask` and `MockDataTask` conform to. That would also mean you wouldn't need to do as much refactoring.

Reply:
Hmm, fair enough. I'll look at doing that instead.